### ISSUE TITLE
[Feature request] Dynamically add new machines in distributed TensorFlow

### ISSUE THREAD

**Issue by qinglintian** [tensorflow/tensorflow#11896]
I'm not sure this has been raised before. I did some search on Google and haven't found relevant stuff. If it do exist, please direct me there. Thank you.

I'm currently experimenting with distributed TensorFlow. When building a distributed cluster, all machines in the cluster should be fed into tf.train.Server as parameters. That is, the disitributed cluster configuration is defined when building the computation graph. Like the example provided in https://github.com/tensorflow/models/blob/master/inception/inception/imagenet_distributed_train.py.

But I have also read papers about robust distributed cluster that it would be nice if the framework support dynamically adding or removing machines if the cluster get larger or some machine goes down.

Is this doable in current version of TensorFlow. If so, is there an example to implement this? If not, is there plans for this?

**Comment by yaroslavvb** [comment]
There's some discussion of this [here](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!msg/discuss/Okb2jKqYwmQ/yxyxXem0BwAJ). Yahoo and TensorflowOnHadoop have said in the past that they are working with TF team to add this

**Comment by byronyi** [comment]
I went through a little bit of the discussion. Given the current session implementation does not support dynamic partitioning of the graph, following the design of `MonitoredSession`, one proposed solution could be:

1. Prepare for a scaling by writing a checkpoint to external storage
2. Terminate current session
3. Re-do the partitioning of graph with respect to the new cluster_def
4. Create a new session and recover from the checkpoint

**Comment by aselle** [comment]
@mrry, @saeta, care to comment. Are you working on this, should we mark this contributions welcome. Do you have any ideas on the design in case we do.

**Comment by saeta** [comment]
We now have some of basics in place (i.e. ClusterSpec propagation, and ClusterResolvers) to allow this sort of thing, however we do not have anyone working on this now as far as I know. The latest tentative designs are pretty much identical to what @byronyl described in his comment above.

**Comment by hongjic** [comment]
How about just scaling the workers? The variables don't need to be re-partitioned when a new worker comes in and the worker just need to build a same graph and pin the variables to the already existed parameter servers. Could that be a little easier to implement?

**Comment by yaroslavvb** [comment]
@hongjic for async parameter server architecture this is already possible. IE, training begins as soon as all parameter server shards are up, and at at least one worker is up. As more worker come online, the training throughput increases.

**Comment by qinglintian** [comment]
Thank you the info provided above. It seems that dynamic scaling of the cluster is not supported yet. Is there any clear plan for supporting that?

@yaroslavvb Correct me if my understanding is wrong. What you mentioned for async parameter server is applicable for data parallelism? When more workers come online, the worker simply build same graph and request data to training the model and transfer updates to parameter server. Then, the effectively batchsize of the cluster increases so we can have larger learning rate as described in [Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour](https://research.fb.com/publications/imagenet1kin1h/). Do we have a mechanism for dynamically adjust hyperparamters?

BTW, in [TensorFlow:Large-Scale Machine Learning on Heterogeneous Distributed Systems](https://arxiv.org/pdf/1603.04467.pdf), it also mentioned parameter replication among servers for server fault tolerant. I suppose that is also not supported in current version of TensorFlow?

**Comment by yaroslavvb** [comment]
@qinglintian if you keep your hyper-parameters (ie, learning rate) in shared variable, then you can adjust from any tensorflow client that's part of the cluster. Fault tolerance is achieved by checkpointing and restarting the training on failures. Permanent removal of workers can be handled with backup workers. Permanent removal of parameter servers is more tricky -- you have to repartition your parameter variables

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by yaroslavvb** [comment]
This is being addressed in https://github.com/tensorflow/tensorflow/issues/14232

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by kzhang28** [comment]
Hi @byronyi 

Following your approach: training --->restart training from checkpoint with a different ClusterSpec
> 1 Prepare for a scaling by writing a checkpoint to external storage
 2 Terminate current session
3 Re-do the partitioning of graph with respect to the new cluster_def
4 Create a new session and recover from the checkpoint

 I have experimented with training my model using synchronous between-graph replicas. After terminating the training process(say 2ps + 8 workers), the training process can be restarted from the last checkpoint and proceed with a different number of worker (say 2ps+10workers).  I was wondering if it is possible to adjust the number of ps . If yes, could you illustrate how to scale the parameter server nodes ? Thank you!

**Comment by umialpha** [comment]
Hi @kzhang28,
Have you test the time cost of restarting the cluster? We are currently schedule tensorflow clusters on k8s. I doubt whether it is worth doing so if we auto scale too often.
